### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by lexfor**
added wave  for CPP

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Comment (KovalM) [inline .idea/.gitignore]: add .idea folder to gitignore

File: .idea/.name [hunk 2]
```diff
@@ -0,0 +1 @@
+wave
```
Comment (KovalM) [inline .idea/.name]: Remove .idea folder from pull request

File: CMakeLists.txt [hunk 3]
```diff
@@ -29,16 +29,16 @@ if (${CMAKE_SYSTEM_NAME} STREQUAL Windows)
...
-	if(BORLAND) 
-		message(STATUS "Compiler: BCC") 
-	endif(BORLAND) 
-	else(${CMAKE_SYSTEM_NAME} STREQUAL Linux) 
-		message(STATUS "Only GCC is supported on Linux")
-	endif(${CMAKE_SYSTEM_NAME} STREQUAL Windows) 
+	if(BORLAND)
+		message(STATUS "Compiler: BCC")
+	endif(BORLAND)
+else(${CMAKE_SYSTEM_NAME} STREQUAL Linux)
```
Comment (KovalM) [inline CMakeLists.txt]: Revert previous formatting in whole file

File: cmake-build-debug/CMakeFiles/3.17.3/CMakeCCompiler.cmake [hunk 4]
```diff
@@ -0,0 +1,76 @@
+set(CMAKE_C_COMPILER "/usr/bin/cc")
```
Comment (KovalM) [inline cmake-build-debug/CMakeFiles/3.17.3/CMakeCCompiler.cmake]: add cmake-build-debug folder to .gitignore and remove it from this pr

File: utils.cpp [hunk 5]
```diff
@@ -1,161 +1,31 @@
...
-        SC_ADDR_MAKE_EMPTY(link);
-        sc_stream_free(stream);
-        return link;
+        std::cout << ")";
     }
-    free(content);
-    sc_stream_free(stream);
-    return link;
-}
+}
```
Comment (KovalM) [inline utils.cpp]: Fix this github warning

File: utils.h [hunk 6]
```diff
@@ -1,31 +1,15 @@
...
- * Creates sc-link with given integer content in sc-memory
-*/
-sc_addr genIntNode(sc_memory_context *context, sc_uint32 data);
-/*
- * Creates sc-link with given float content in sc-memory
-*/
-sc_addr genFloatNode(sc_memory_context *context, float data);
 
-#endif // UTILS_H
+#endif // UTILS_H
```
Comment (KovalM) [inline utils.h]: Remove this github warning

File: utils.h [hunk 7]
```diff
@@ -1,31 +1,15 @@
...
  * Prints identifier of given sc-element.
  * If element has not identifier, prints its address in sc-memory
  * If element is arc prints it in (a -> b) form recursively.
 */
-void printEl(sc_memory_context *context, sc_addr element);
+void printEl(const std::unique_ptr<ScMemoryContext>& context, ScAddr element);
 /*
  * Prints content of given sc-link
 */
-void printContent(sc_memory_context *context, sc_addr element);
```
Comment (KovalM) [inline utils.h:13]: Why we remove all these methods?

File: wavefindpath.cpp [hunk 8]
```diff
@@ -11,142 +6,110 @@ extern "C" {
...
 #include <algorithm>
+#include "cpp/sc_addr.hpp"
+#include "cpp/sc_memory.hpp"
+#include "cpp/sc_iterator.hpp"
+#include "utils.h"
 
-using namespace std;
-
-sc_memory_context *context;
+//sc_memory_context *context;
```
Comment (KovalM) [inline wavefindpath.cpp]: Remove it if we don't use

**Review by KovalM — CHANGES_REQUESTED** [review]
Add Readme with getting started

File: utils.h (hunk 7 above)
Comment (lexfor) [reply utils.h:13]: they aren't use in program

**Comment by lexfor** [comment]
all warings fixed, excess folders delete and readme added

File: wavefindpath.cpp [hunk 9]
```diff
@@ -155,233 +111,189 @@ sc_addr get_other_vertex_incidence_edge(sc_addr edge, sc_addr vertex)
...
+
 int main()
 {
     sc_memory_params params;
 
     sc_memory_params_clear(&params);
-    params.repo_path = "../../kb.bin";
-    params.config_file = "../../config/sc-web.ini";
-    params.ext_path = "./extensions";
+    params.repo_path = "/home/tima/ostis/kb.bin";
```
Comment (KovalM) [inline wavefindpath.cpp]: Use relative path.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,288 @@
+                                                          КП ППвИС(часть 1)
+                                                             Установка
+
+Для выполнения первой части КП по курсу ППвИС разработан пример реализации волнового алгоритма поиска минимального пути в графе на базе библиотеки sc-memory.
+
+На вашем компьютере должна быть установлена версия ostis ветки 0.5.0 из репозитория: https://github.com/ShunkevichDV/ostis и Clion.
```
Comment (KovalM) [inline README.md]: Use this link https://github.com/ostis-dev/ostis-web-platform. Update all file with ostis-web-platform insted of ostis

File: README.md [hunk 11]
```diff
@@ -0,0 +1,288 @@
+                                                          КП ППвИС(часть 1)
```
Comment (KovalM) [inline README.md]: Use markdown styles for format file. Readme file should be readable
image

File: README.md (hunk 10 above)
Comment (KovalM) [inline README.md]: Clion used as example but it is not necessarily

File: README.md [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+1) Запустите терминал и перейдите в папку ostis, выполните команду git clone https://github.com/ostis-apps/wave_find_path_sc_memory
+
+2) Перейдите в папку wave_find_path_sc_memory и скопируйте папку graph в ostis/kb
+
+3) Запустите Clion. Запустив среду, необходимо выбрать пункт меню Open Project и в появившемся диалоговом окне указать файл CMakeLists.txt,
+находящийся в папке с примером.
+
+4) После этого проект будет виден во вкладке Редактор(Edit). Перейдите в файл wavefindpath.cpp
+
+5) Измените переменные в функции main
```
Comment (KovalM) [inline README.md:25]: Can we use relative path and drop this point?

File: README.md [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+    params.repo_path = "/home/<имя_пользователя>/../kb.bin";
+    params.config_file = "/home/<имя_пользователя>/../config/sc-web.ini";
+    params.ext_path = "/home/<имя_пользователя>/../sc-machine/bin/extensions";
+
+6) Перейдите в файл /home/<имя_компьютера>/../config/sc-web.ini и измените в нем строчки Path и Directory на:
+    Path = /home/<имя_пользователя>/../kb.bin
+    Directory = /home/<имя_пользователя>/../sc-machine/bin/extensions
+
+7) Пересоберите базу знаний (~/../scripts$ ./build_kb.sh)
```
Comment (KovalM) [inline README.md]: ~/../scripts$ what is it?

File: CMakeLists.txt (hunk 3 above)
Comment (KovalM) [reply CMakeLists.txt]: not fixed

**Review by KovalM — CHANGES_REQUESTED** [review]

File: README.md (hunk 12 above)
Comment (lexfor) [reply README.md:25]: no, need to use absolut path

File: README.md [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+Для выполнения первой части КП по курсу ППвИС разработан пример реализации волнового алгоритма поиска минимального пути в графе на базе библиотеки sc-memory.
+
+На вашем компьютере должна быть установлена версия ostis ветки 0.5.0 из репозитория: https://github.com/ostis-dev/ostis-web-platform.
+
+Последние версии указанного примера и данного документа могут быть найдены в репозитории: https://github.com/ostis-apps/wave_find_path_sc_memory
+или на сервере: Info/Studinfo/~Методическое обеспечение кафедры/~Учебные курсы/2 курс/ПпвИС/1sem/КП 2020
+
+1) Запустите терминал и перейдите в папку ostis-web-platform, выполните команду git clone https://github.com/ostis-apps/wave_find_path_sc_memory
+
+2) Перейдите в папку wave_find_path_sc_memory и скопируйте папку graph в ostis-web-platform/kb
```
Comment (KovalM) [inline README.md]:
```suggestion
2) Откройте файл ostis-web-platform/repo.path и допишите в него строчку: wave_find_path_sc_memory/graph
```

**Review by KovalM — CHANGES_REQUESTED** [review]

**Pull request merged**